python/ctranslate2/converters/transformers.py

```diff
@@ -627,6 +627,98 @@ def set_decoder(self, spec, module, rotary_dim, num_heads):
             self.set_linear(layer_spec.ffn.linear_1, layer.mlp.fc_out)
 
 
+@register_loader("CodeGenConfig")
+class CodeGenLoader(ModelLoader):
+    @property
+    def architecture_name(self):
+        return "CodeGenForCausalLM"
+
+    def get_model_spec(self, model):
+        spec = transformer_spec.TransformerDecoderModelSpec.from_config(
+            model.config.n_layer,
+            model.config.n_head,
+            pre_norm=True,
+            activation=_SUPPORTED_ACTIVATIONS[model.config.activation_function],
+            rotary_dim=model.config.rotary_dim,
+            rotary_interleave=False,
+            parallel_residual=True,
+            shared_layer_norm=True,
+        )
+
+        mp_num = 4
+        if (
+            "Salesforce/codegen2-1B" in model.name_or_path
+            or "Salesforce/codegen2-3_7B" in model.name_or_path
+        ):
+            # consider mp_num=8 for smaller codegen2 series.
+            mp_num = 8
+
+        self.set_decoder(
+            spec.decoder,
+            model.transformer,
+            model.config.rotary_dim,
+            model.config.n_head,
+            model.config.n_embd,
+            mp_num=mp_num,
+        )
+        self.set_linear(spec.decoder.projection, model.lm_head)
+        return spec
+
+    def get_vocabulary(self, model, tokenizer):
+        tokens = super().get_vocabulary(model, tokenizer)
+
+        extra_ids = model.config.vocab_size - len(tokens)
+        for i in range(extra_ids):
+            # fix for additional vocab, see GPTNeoX Converter
+            tokens.append("<extra_id_%d>" % i)
+
+        return tokens
+
+    def set_vocabulary(self, spec, tokens):
+        spec.register_vocabulary(tokens)
+
+    def set_config(self, config, model, tokenizer):
+        config.bos_token = tokenizer.bos_token
+        config.eos_token = tokenizer.eos_token
+        config.unk_token = tokenizer.unk_token
+
+    def set_decoder(self, spec, module, rotary_dim, num_heads, embed_dim, mp_num):
+        spec.scale_embeddings = False
+        self.set_embeddings(spec.embeddings, module.wte)
+        self.set_layer_norm(spec.layer_norm, module.ln_f)
+
+        _base_permutation = np.arange(0, mp_num * 3).reshape(-1, 3).T.flatten().tolist()
+        local_dim = embed_dim // mp_num
+        permutation = np.concatenate(
+            [np.arange(i * local_dim, (i + 1) * local_dim) for i in _base_permutation]
+        )
```

Comment:
No need for a leading underscore here.

```suggestion
        base_permutation = np.arange(0, mp_num * 3).reshape(-1, 3).T.flatten().tolist()
        local_dim = embed_dim // mp_num
        permutation = np.concatenate(
            [np.arange(i * local_dim, (i + 1) * local_dim) for i in base_permutation]
        )
```